File: frontend/src/lib/clientConfig.ts
/** Cadences the server decides and the client runs at.

The drawer's flush interval is the one this exists for. It is the largest
single lever on drawing bandwidth, and the value that turned out to be right
was not the one the byte curve pointed at — it was found by looking at a
viewer's screen, and a value only looking can settle is one somebody has to be
able to change while looking. So the server ships it rather than the bundle
compiling it.

Read where the socket lives rather than in a component, for the same reason
`server_full` is: the notice arrives at the handshake, which may be long before
anything that cares has mounted, and a value that arrived early must not be
lost. Subscribers are therefore called immediately with what is already known.

A missing or malformed field falls back to the compiled default. A server that
cannot say is not a reason to stop drawing, and these numbers all have an
answer that has always worked.

Version 2 dropped `lobbyPollIntervalMs`: the lobby is told about rooms over its
channel now (#462) and has no cadence of its own to be given. Version 3 added
the drawing allowance (#597). Version 4 added the AFK input window (#677),
which is the same kind of number for the same reason: only watching somebody
play settles how recently a client must have seen a hand on the controls
before it answers an AFK check on their behalf. */

export interface ClientConfig {
  flushIntervalMs: number;
  /** The drawing allowance a `draw` frame spends: this many frames per
  window. Version 3 (#597) added it so a replay after a stall can pace itself
  under the allowance instead of bursting into a silent refusal. */
  drawingFramesPerWindow: number;
  drawingWindowSeconds: number;
  /** How recently this client must have seen a pointer or a key to answer an
  `afk_check` for the player, rather than asking them. Version 4 (#677). */
  afkInputWindowMs: number;
}

/** What the client uses until a server says otherwise, and if one never does. */
export const DEFAULT_CLIENT_CONFIG: ClientConfig = {
  flushIntervalMs: 80,
  drawingFramesPerWindow: 100,
  drawingWindowSeconds: 2,
  afkInputWindowMs: 60_000,
};

/** The bounds the server enforces, mirrored so a bad payload cannot get through.

Not a second opinion about the right value — the server owns that. This only
refuses a number that would break the client outright: a zero or negative
interval is a busy loop, and a huge one is a canvas that never updates. */
const BOUNDS: Record<keyof ClientConfig, { min: number; max: number }> = {
  flushIntervalMs: { min: 10, max: 200 },
  drawingFramesPerWindow: { min: 50, max: 400 },
  drawingWindowSeconds: { min: 0.5, max: 10 },
  // A window of zero would ask somebody who is typing; one of an hour would
  // answer for somebody who left. Neither breaks the client, but both defeat
  // the check, so the same mirrored-bounds rule applies.
  afkInputWindowMs: { min: 5_000, max: 600_000 },
};

function reading(
  payload: Record<string, unknown> | undefined,
  key: keyof ClientConfig,
): number {
  const value = payload?.[key];
  if (typeof value !== "number" || !Number.isFinite(value)) {
    return DEFAULT_CLIENT_CONFIG[key];
  }
  const { min, max } = BOUNDS[key];
  if (value < min || value > max) return DEFAULT_CLIENT_CONFIG[key];
  return value;
}

/** The notice shape this build understands. */
export const CLIENT_CONFIG_CONTRACT_VERSION = 4;

/** Read a `client_config` notice, or `null` if it is not one this build knows.

The version is checked as a whole before any field is read, the same way the
shutdown and pause notices are. A later server could give a field a different
meaning rather than a different name — a window instead of an interval, say —
and a client that took the fields it recognised would apply half of a contract
it does not understand. Ignoring the notice leaves the cadences that have
always worked, which is the safe direction and the one every field-level
fallback below already takes. */
export function parseClientConfig(payload: unknown): ClientConfig | null {
  if (typeof payload !== "object" || payload === null) return null;
  const record = payload as Record<string, unknown>;
  if (record.contractVersion !== CLIENT_CONFIG_CONTRACT_VERSION) return null;
  return {
    flushIntervalMs: reading(record, "flushIntervalMs"),
    drawingFramesPerWindow: reading(record, "drawingFramesPerWindow"),
    drawingWindowSeconds: reading(record, "drawingWindowSeconds"),
    afkInputWindowMs: reading(record, "afkInputWindowMs"),
  };
}

let current: ClientConfig = DEFAULT_CLIENT_CONFIG;
const listeners = new Set<(config: ClientConfig) => void>();

/** The cadences in force right now. */
export function currentClientConfig(): ClientConfig {
  return current;
}

/** Adopt a notice from the server, telling everyone who is listening.

Silent when nothing moved, so a reconnect — which re-sends the same values —
does not tear down and re-arm every timer that depends on them. */
export function applyClientConfig(payload: unknown): ClientConfig {
  const next = parseClientConfig(payload);
  if (next === null) return current;
  const unchanged = (
    Object.keys(next) as (keyof ClientConfig)[]
  ).every((key) => next[key] === current[key]);
  if (unchanged) return current;
  current = next;
  listeners.forEach((listener) => listener(current));
  return current;
}

/** Subscribe to the cadences. Called immediately with what is known already. */
export function onClientConfig(
  listener: (config: ClientConfig) => void,
): () => void {
  listeners.add(listener);
  listener(current);
  return () => {
    listeners.delete(listener);
  };
}

/** Forget everything, for tests that need a clean channel. */
export function resetClientConfig(): void {
  current = DEFAULT_CLIENT_CONFIG;
  listeners.clear();
}

File: backend/app/client_config.py
"""Client-side cadences the server decides, and the notice that carries them.

Most tunables change something the server does. This one changes something the
*client* does, which is the harder half of #446 and the half the issue was
actually about: the drawer's flush interval is the largest single lever on
drawing bandwidth, and the value that turned out to be right was not the one
the byte curve pointed at. It was found by looking at a viewer's screen, and a
value that can only be found by looking is a value somebody has to be able to
change while looking.

So they are shipped rather than compiled. The carrier is a notice sent to each
socket at the handshake and re-sent to everyone when a value changes — not
`room_state`, which is per-room and never reaches somebody sitting in the
lobby, and not the acknowledgement, which the handshake does not have.

`contractVersion` follows `server_shutdown`: this payload has a shape of its
own that can change without the whole protocol moving.
"""
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.handlers.budgets import Budget

CLIENT_CONFIG_CONTRACT_VERSION = 4


def _compiled_drawing_budget() -> Budget:
    # Imported here: the handlers package imports this module at load time.
    from app.handlers.budgets import DRAWING

    return DRAWING.default


@dataclass
class ClientConfig:
    """The cadences this server is asking its clients to run at."""

    # How long queued path points wait before going out as one frame. The
    # drawer never feels it - their own canvas is rasterized on every
    # pointermove. A viewer used to paint each batch the moment it landed,
    # which made 56 ms and 80 ms read as steppy on a viewer's screen and kept
    # this at 40; since #559 a viewer plays each batch out over the interval
    # that follows it, at the screen's own rate, so the interval no longer
    # shows as steps and the byte curve gets its way: 80 ms halves the
    # point messages a drawer sends, at the cost of a viewer seeing ink up
    # to 80 ms behind the drawer's hand instead of 40. Still shipped, so it
    # can be moved back from the admin panel while somebody watches.
    flush_interval_ms: int = 80

    # Where the drawing budget in force is read from. Version 3 (#597) tells
    # the client the allowance its `draw` frames spend, so a client replaying
    # a stroke after a stall can pace itself under it instead of bursting into
    # a refusal that drops the frame nobody is waiting on. It is a callable,
    # not a copy: the budget is a tunable, and the notice is re-sent when it
    # moves (`announce_client_config`), so the value read must be the live one.
    drawing_budget: Callable[[], Budget] = field(default=_compiled_drawing_budget)

    # How recently a client must have seen a pointer or a key to answer an
    # AFK check on the player's behalf (#677). Version 4, and it belongs here
    # for the reason the module exists: it changes something the *client*
    # does, and it is the one number in the feature that can only be settled
    # by watching somebody play. Too short and a person who is reading the
    # canvas gets a dialog; too long and a client answers for somebody who
    # left a minute ago. Shipped, so it can be moved while somebody watches.
    afk_input_window_ms: int = 60_000

    def payload(self) -> dict:
        """The `client_config` notice, in the names the client reads."""
        budget = self.drawing_budget()
        return {
            "contractVersion": CLIENT_CONFIG_CONTRACT_VERSION,
            "flushIntervalMs": self.flush_interval_ms,
            "drawingFramesPerWindow": budget.limit,
            "drawingWindowSeconds": budget.window_seconds,
            "afkInputWindowMs": self.afk_input_window_ms,
        }


# One process, one answer for every client that connects to it.
client_config = ClientConfig()
